sdks/python/apache_beam/runners/portability/artifact_service.py

```diff
@@ -263,3 +279,205 @@ def _open(self, path, mode='r'):
+    ):
+    self._lock = threading.Lock()
+    self._jobs_to_stage = {}
+    self._file_writer = file_writer
+
+  def register_job(self, staging_token, dependencies):
+    self._jobs_to_stage[staging_token] = list(dependencies), threading.Event()
+
+  def resolved_deps(self, staging_token, timeout=None):
+    dependencies_list, event = self._jobs_to_stage[staging_token]
+    try:
+      if not event.wait(timeout):
+        raise concurrent.futures.TimeoutError()
+      return dependencies_list
+    finally:
+      del self._jobs_to_stage[staging_token]
+
+  def ReverseArtifactRetrievalService(self, responses, context=None):
+    staging_token = next(responses).staging_token
+    dependencies, event = self._jobs_to_stage[staging_token]
+
+    requests = queue.Queue()
+
+    class FakeRetrievalService(object):
+      def ResolveArtifacts(self, request):
+        requests.put(
+            beam_artifact_api_pb2.ArtifactRequestWrapper(
+                resolve_artifact=request))
+        return next(responses).resolve_artifact_response
+
+      def GetArtifact(self, request):
+        requests.put(
+            beam_artifact_api_pb2.ArtifactRequestWrapper(get_artifact=request))
+        while True:
+          response = next(responses)
+          yield response.get_artifact_response
+          if response.is_last:
+            break
+
+    def resolve():
+      file_deps = resolve_as_files(
+          FakeRetrievalService(),
+          lambda name: self._file_writer(os.path.join(staging_token, name)),
+          dependencies)
+      dependencies[:] = file_deps
+      requests.put(None)
+      event.set()
+
+    t = threading.Thread(target=resolve)
+    t.daemon = True
+    t.start()
+
+    return _queue_iter(requests, None)
+
+
+def resolve_as_files(retrieval_service, file_writer, dependencies):
+  """Translates a set of dependencies into file-based dependencies."""
+  # Resolve until nothing changes.  This ensures that they can be fetched.
+  resolution = retrieval_service.ResolveArtifacts(
+      beam_artifact_api_pb2.ResolveArtifactRequest(
+          artifacts=dependencies,
+          # Anything fetchable will do.
+          # TODO(robertwb): Take advantage of shared filesystems, urls.
+          preferred_urns=[],
+      ))
+  if resolution.error:
+    raise RuntimeError(resolution)
+  dependencies = resolution.replacements
+
+  # Fetch each of the dependencies, using file_writer to store them as
+  # file-based artifacts.
+  # TODO(robertwb): Consider parallelizing the actual writes.
+  for dep in dependencies:
+    if dep.role_urn == common_urns.artifact_roles.STAGING_TO.urn:
+      base_name = os.path.basename(
+          proto_utils.parse_Bytes(
+              dep.role_payload,
+              beam_runner_api_pb2.ArtifactStagingToRolePayload).staged_name)
+    else:
+      base_name = None
+    unique_name = '-'.join(
+        filter(
+            None,
+            [hashlib.sha256(dep.SerializeToString()).hexdigest(), base_name]))
```

Comment:
The full file path will already be prefixed by the staging token (line ~413 where this is called). This does make me wonder whether every artifact should have a name (even non-file based ones). Especially if roles become more specific (like classpath or pip package rather than just stage_to).

Reply:
Its typically pretty easy to figure out the name of a path component so no need.